File: eeg-backend/neuro_yogic/iaf.py
"""
iaf.py
======
Individual Alpha Frequency — the reference oracle for the C# port in
``src/NeuroYogic.SignalProcessing/Dsp/Iaf.cs``.

The alpha peak sits at different frequencies for different people (~8–12 Hz), and
the multi-session literature found that *generalised* bands measurably hurt
classification — individual band-centering helps. We estimate the IAF as the
spectral centre-of-gravity of power in 7–13 Hz (more stable than a raw argmax),
per channel then averaged. Phase 1 computes the scalar only; per-user band
re-centering is deferred to Phase 4c.
"""
import numpy as np


def individual_alpha_freq(freqs, psd, lo: float = 7.0, hi: float = 13.0) -> float:
    """Alpha centre-of-gravity (Hz) over [lo, hi], averaged across channels.

    Returns 0.0 if there is no alpha-band power.
    """
    freqs = np.asarray(freqs, dtype=np.float64)
    psd = np.atleast_2d(np.asarray(psd, dtype=np.float64))
    mask = (freqs >= lo) & (freqs <= hi)
    f = freqs[mask]
    if f.size == 0:
        return 0.0

    peaks = []
    for ch in psd:
        p = ch[mask]
        denom = float(np.sum(p))
        if denom <= 0.0:
            continue
        peaks.append(float(np.sum(f * p) / denom))
    return float(np.mean(peaks)) if peaks else 0.0

File: eeg-backend/src/NeuroYogic.SignalProcessing/Dsp/Iaf.cs
namespace NeuroYogic.SignalProcessing.Dsp;

/// <summary>
/// Individual Alpha Frequency — faithful C# port of the Python oracle in
/// <c>neuro_yogic/iaf.py</c>. Estimates the alpha peak as the spectral
/// centre-of-gravity of power in 7–13 Hz (more stable than a raw argmax), per
/// channel then averaged.
/// </summary>
internal static class Iaf
{
    public static double Estimate(double[] freqs, double[][] psd, double lo = 7.0, double hi = 13.0)
    {
        var maskIdx = new List<int>();
        for (var i = 0; i < freqs.Length; i++)
            if (freqs[i] >= lo && freqs[i] <= hi) maskIdx.Add(i);
        if (maskIdx.Count == 0) return 0.0;

        double sum = 0.0;
        var count = 0;
        foreach (var ch in psd)
        {
            double num = 0.0, den = 0.0;
            foreach (var i in maskIdx) { num += freqs[i] * ch[i]; den += ch[i]; }
            if (den <= 0.0) continue;
            sum += num / den;
            count++;
        }
        return count > 0 ? sum / count : 0.0;
    }
}
